### PR TITLE
Add update_service func

### PR DESCRIPTION
**Pull request by wangxiyuan**
1. Add update_service func which will be used by keepalived. Note that this is not for operators so that we won't expose cli.

2. Fixed the bug that service's updated_at is always None.

3. Fixed the bug that the service's node role is always the same.

4. Fixed the bug that `node set` cli doesn't work.

5. Add `node_name_filter` and `node_role_filter` for `list_services` function

6. Expose `node_name` for service cmds.

### CODE REVIEW

File: openlabcmd/openlabcmd/zk.py [hunk 1]
```diff
@@ -206,37 +207,100 @@ def delete_node(self, node_name):
...
             for service_name in self.client.get_children(path):
                 service_path = path + '/' + service_name
                 service_info = self.client.get(service_path)
+                mtime = service_info[1].mtime
                 service_info = self._bytesToDict(service_info[0])
+                service_info['updated_at'] = datetime.datetime.fromtimestamp(
+                    mtime / 1000).isoformat()
+                service_info['node_name'] = exist_node['name']
                 result.append(service_info)
         return sorted(result)
```
Comment (kiwik) [inline openlabcmd/openlabcmd/zk.py]: Sort a zookeeper object list, so what is expected return value? order by `node_name`?

File: openlabcmd/openlabcmd/zk.py [hunk 2]
```diff
@@ -206,37 +207,100 @@ def delete_node(self, node_name):
...
-                exceptions.ClientError("Can't find service %s" % service_name)
+    def _get_service_role(self, service_name):
+        if (service_name in
+                service.service_mapping['master_service']['necessary'] +
+                service.service_mapping['master_service']['unnecessary']):
+            return 'master'
+        elif (service_name in
+              service.service_mapping['slave_service']['necessary'] +
+              service.service_mapping['slave_service']['unnecessary']):
+            return 'slave'
```
Comment (kiwik) [inline openlabcmd/openlabcmd/zk.py:274]: Please add some comments about this logic, I originally thought I should add the following comments until I found the logic in cli.py :)

> we can't use service_name to decide service role, like: `mysql` and `zookeeper`, they are both in master and slave nodes. Imagine we input `mysql` into the function, always return `master`

File: openlabcmd/openlabcmd/zk.py [hunk 3]
```diff
@@ -206,37 +207,100 @@ def delete_node(self, node_name):
...
+
+        exceptions.ClientError("Can't find service %s" % service_name)
```
Comment (kiwik) [inline openlabcmd/openlabcmd/zk.py]: Raise?

File: openlabcmd/openlabcmd/zk.py [hunk 4]
```diff
@@ -206,37 +207,114 @@ def delete_node(self, node_name):
...
+            return 'slave'
 
+        exceptions.ClientError("Can't find service %s" % service_name)
+
+    def _get_service_path_and_node(self, service_name, role):
         for exist_node in self.list_nodes():
             if exist_node['role'] == role:
                 path = '/ha/%s/%s_service/%s' % (exist_node['name'], role,
                                                  service_name)
-                result = self.client.get(path)
```
Comment (kiwik) [inline openlabcmd/openlabcmd/zk.py:239]: Why do not check the service existed? if service_name=`not_existed_service`?

File: openlabcmd/openlabcmd/zk.py [hunk 5]
```diff
@@ -206,37 +207,114 @@ def delete_node(self, node_name):
...
+    def get_service(self, service_name, role=None):
+        if not role:
+            role = self._get_service_role(service_name)
+        path, srv_node = self._get_service_path_and_node(service_name, role)
+        result = self.client.get(path)
+        mtime = result[1].mtime
+        result = self._bytesToDict(result[0])
+        result['updated_at'] = datetime.datetime.fromtimestamp(
+            mtime / 1000).isoformat()
+        result['node_name'] = srv_node['name']
```
Comment (kiwik) [inline openlabcmd/openlabcmd/zk.py:293]: I see line#277~#281 logic multiple times in different places, please add a function to handle this.

File: openlabcmd/openlabcmd/service.py [hunk 6]
```diff
@@ -26,27 +26,31 @@ class ServiceStatus(object):
     INITIALIZING = 'initializing'
     UP = 'up'
     DOWN = 'down'
-    Restarting = 'restarting'
+    RESTARTING = 'restarting'
+
+    @property
+    def all_services(self):
```
Comment (kiwik) [inline openlabcmd/openlabcmd/service.py]: s/all_services/all_status/

File: openlabcmd/openlabcmd/zk.py [hunk 7]
```diff
@@ -206,37 +207,114 @@ def delete_node(self, node_name):
...
+                raise exceptions.ValidationError('restarted should be '
+                                                 'boolean.')
+            old_service['restarted'] = restarted
+            old_service['restarted_at'] = current_time
+        if status:
+            if status not in service.ServiceStatus.all_services:
+                raise exceptions.ValidationError(
+                    'status should be in %s.' %
+                    service.ServiceStatus.all_services)
+            old_service['status'] = status
```
Comment (kiwik) [inline openlabcmd/openlabcmd/zk.py:326]: Don't forget to update `updated_at` attribute.

File: openlabcmd/openlabcmd/zk.py [hunk 8]
```diff
@@ -206,37 +207,114 @@ def delete_node(self, node_name):
...
+        if not role:
+            if service_name in service.MIXED_SERVICE:
+                raise exceptions.ValidationError(
+                    "Role must be specified if service is in "
+                    "%s." % service.MIXED_SERVICE)
+            role = self._get_service_role(service_name)
+        path, _ = self._get_service_path_and_node(service_name, role)
+        current_time = datetime.datetime.now().isoformat()
+
+        if alarmed:
```
Comment (kiwik) [inline openlabcmd/openlabcmd/zk.py]: How to update `alarmed` to false?

File: openlabcmd/openlabcmd/zk.py [hunk 9]
```diff
@@ -206,37 +207,114 @@ def delete_node(self, node_name):
...
+            if not isinstance(alarmed, bool):
+                raise exceptions.ValidationError('alarmed should be boolean.')
+            old_service['alarmed'] = alarmed
+            old_service['alarmed_at'] = current_time
+        if restarted:
```
Comment (kiwik) [inline openlabcmd/openlabcmd/zk.py]: ditto

**Review by kiwik — CHANGES_REQUESTED** [review]

File: openlabcmd/openlabcmd/zk.py (hunk 7 above)
Comment (wangxiyuan) [reply openlabcmd/openlabcmd/zk.py:326]: `updated_at` is handled by zookeeper automatically. Every node in zookeeper has some build-in properties like ctime (created_at), mtime(updated_at), once the node info is changed, `mtime` will be changed as well.

File: openlabcmd/openlabcmd/zk.py (hunk 8 above)
Comment (wangxiyuan) [reply openlabcmd/openlabcmd/zk.py]: oh, my fault here. It should be `if alarmed is not None`

File: openlabcmd/openlabcmd/zk.py (hunk 5 above)
Comment (wangxiyuan) [reply openlabcmd/openlabcmd/zk.py:293]: I'll merge them into `_bytesToDict` func

File: openlabcmd/openlabcmd/zk.py (hunk 4 above)
Comment (wangxiyuan) [reply openlabcmd/openlabcmd/zk.py:239]: This function is only used for generate a path which will be used later, if the service doesn't exist, it'll raise NotFound error later.

File: openlabcmd/openlabcmd/zk.py (hunk 3 above)
Comment (wangxiyuan) [reply openlabcmd/openlabcmd/zk.py]: ++

File: openlabcmd/openlabcmd/zk.py (hunk 2 above)
Comment (wangxiyuan) [reply openlabcmd/openlabcmd/zk.py:274]: OK. will do

File: openlabcmd/openlabcmd/zk.py (hunk 1 above)
Comment (wangxiyuan) [reply openlabcmd/openlabcmd/zk.py]: Done

File: openlabcmd/openlabcmd/service.py (hunk 6 above)
Comment (wangxiyuan) [reply openlabcmd/openlabcmd/service.py]: Done

File: openlabcmd/openlabcmd/zk.py (hunk 4 above)
Comment (kiwik) [reply openlabcmd/openlabcmd/zk.py:239]: OK, thanks for your clarification

File: openlabcmd/openlabcmd/zk.py (hunk 7 above)
Comment (kiwik) [reply openlabcmd/openlabcmd/zk.py:326]: I see the column `updated_at` https://github.com/theopenlab/labkeeper/pull/36/files#diff-fe16cf954dd7bc0d309b1d1d1f2e1441R39 , that will be mapped to `mtime` automatically?

Comment (wangxiyuan) [reply openlabcmd/openlabcmd/zk.py:326]: Yes, see L50, I mapped `mtime` to `updated_at`  there when fetching data from zk.

File: openlabcmd/openlabcmd/cli.py [hunk 10]
```diff
@@ -124,6 +124,11 @@ def _add_ha_service_cmd(self, parser):
...
             'list', help='List all services.')
         cmd_ha_service_list.set_defaults(func=self.ha_service_list)
+        cmd_ha_service_list.add_argument(
+            '--node', help='Filter the services with the specified node.')
+        cmd_ha_service_list.add_argument(
+            '--role', choices=['master', 'slave', 'zookeeper'],
+            help='Filter the services with the specified node role.')
         # openlab ha service get
         cmd_ha_service_get = cmd_ha_service_subparsers.add_parser(
             'get', help='Get a services.')
```
Comment (huangtianhua) [inline openlabcmd/openlabcmd/cli.py]: Not relate with this pr: Get a service?

Comment (wangxiyuan) [reply openlabcmd/openlabcmd/cli.py]: Please take a look at the pr commit message. I did a lot change in this pr.

File: openlabcmd/openlabcmd/node.py [hunk 11]
```diff
@@ -10,17 +10,14 @@ class NodeStatus(object):
...
         self.ip = ip
-        if not heartbeat:
-            self.heartbeat = 0
-        if not alarmed:
-            self.alarmed = False
-        if not status:
-            self.status = NodeStatus.INITIALIZING
+        self.heartbeat = heartbeat if heartbeat else 0
+        self.alarmed = alarmed if alarmed else False
+        self.status = status if status else NodeStatus.INITIALIZING
```
Comment (huangtianhua) [inline openlabcmd/openlabcmd/node.py]: Why not set the default values for these parameters, then we can set them simply?

File: openlabcmd/openlabcmd/service.py [hunk 12]
```diff
@@ -26,27 +26,31 @@ class ServiceStatus(object):
...
-        self.restarted_at = None
-        self.is_necessary = None
         self.node_role = node_role
-        self.status = ServiceStatus.INITIALIZING
+        self.alarmed = alarmed if alarmed else False
+        self.alarmed_at = alarmed_at if alarmed_at else None
+        self.restarted = restarted if restarted else False
+        self.restarted_at = restarted_at if restarted_at else None
+        self.is_necessary = is_necessary if is_necessary else None
+        self.status = status if status else ServiceStatus.INITIALIZING
```
Comment (huangtianhua) [inline openlabcmd/openlabcmd/service.py]: same.

File: openlabcmd/openlabcmd/node.py (hunk 11 above)
Comment (wangxiyuan) [reply openlabcmd/openlabcmd/node.py]: It works here.

But it is not a good custom for python developer IMO. It works terribly for inherit class or dict/list args. Here is an article explain clearly: https://graysonkoonce.com/always-use-none-for-default-args-in-python/

I'm used to using `None` for default value.

File: openlabcmd/openlabcmd/node.py [hunk 13]
```diff
@@ -10,17 +10,14 @@ class NodeStatus(object):
...
         self.role = role
         self.type = type
         self.ip = ip
-        if not heartbeat:
-            self.heartbeat = 0
-        if not alarmed:
-            self.alarmed = False
-        if not status:
-            self.status = NodeStatus.INITIALIZING
+        self.heartbeat = heartbeat if heartbeat else 0
```
Comment (liusheng) [inline openlabcmd/openlabcmd/node.py]: `= heartbeat or 0`

Comment (liusheng) [reply openlabcmd/openlabcmd/node.py]: can the `heartbeat` be initialized with 0 ? that may cause `if heartbeat` clause to be `False`

File: openlabcmd/openlabcmd/node.py [hunk 14]
```diff
@@ -10,17 +10,14 @@ class NodeStatus(object):
...
+        self.alarmed = alarmed if alarmed else False
```
Comment (liusheng) [inline openlabcmd/openlabcmd/node.py]: alarmed or False

File: openlabcmd/openlabcmd/node.py [hunk 15]
```diff
@@ -34,7 +31,6 @@ def to_zk_data(self):
         }
         return json.dumps(node_dict).encode('utf8')
 
-    @classmethod
-    def from_dict(cls, name, role, n_type, ip, heartbeat=None, alarmed=None,
-                  status=None):
-        return cls(name, role, n_type, ip, heartbeat, alarmed, status)
+    @ classmethod
+    def from_dict(cls, kwargs):
```
Comment (liusheng) [inline openlabcmd/openlabcmd/node.py]: not sure if `kwargs` is a good  input variable name. generally, it is used to be refer to `input keyword arguments`, but here you need to input a dict, right ? not similar to the above old function. I am also not very sure.

File: openlabcmd/openlabcmd/service.py [hunk 16]
```diff
@@ -26,27 +26,31 @@ class ServiceStatus(object):
...
-        self.restarted = False
-        self.restarted_at = None
-        self.is_necessary = None
         self.node_role = node_role
-        self.status = ServiceStatus.INITIALIZING
+        self.alarmed = alarmed if alarmed else False
+        self.alarmed_at = alarmed_at if alarmed_at else None
+        self.restarted = restarted if restarted else False
+        self.restarted_at = restarted_at if restarted_at else None
+        self.is_necessary = is_necessary if is_necessary else None
```
Comment (liusheng) [inline openlabcmd/openlabcmd/service.py]: why need `... if ... else...` ? these vars already have the default `None` value

File: openlabcmd/openlabcmd/zk.py [hunk 17]
```diff
@@ -35,8 +35,22 @@ def __init__(self, config=None):
                                          "object.")
         self._last_retry_log = 0
 
-    def _bytesToDict(self, data):
-        return json.loads(data.decode('utf8'))
+    def _bytesToDict(self, node_data, include_created_at=False,
+                     include_updated_at=False):
+        """ Convert zookeeper node data into dict."""
+        node_dict = node_data[0]
```
Comment (liusheng) [inline openlabcmd/openlabcmd/zk.py:41]: node_data is a list and length always is 1 ?

File: openlabcmd/openlabcmd/zk.py [hunk 18]
```diff
@@ -206,37 +217,116 @@ def delete_node(self, node_name):
         self.client.delete(path, recursive=True)
 
     @_client_check_wrapper
-    def list_services(self):
+    def list_services(self, node_name_filter=None, node_role_filter=None):
+        """
+        List the services in the HA deployment.
+        :param node_name_filter: The node filter.
+        :type node_name_filter: list or string.
```
Comment (liusheng) [inline openlabcmd/openlabcmd/zk.py:224]: does the `--role` and `--node` arguments can be parsed to a list currently ?

File: openlabcmd/openlabcmd/zk.py [hunk 19]
```diff
@@ -206,37 +217,116 @@ def delete_node(self, node_name):
...
+                raise exceptions.ValidationError(
+                    "Role must be specified if service is in "
+                    "%s." % service.MIXED_SERVICE)
+            role = self._get_service_role(service_name)
+        path, _ = self._get_service_path_and_node(service_name, role)
+        current_time = datetime.datetime.now().isoformat()
+
+        if alarmed is not None:
+            if not isinstance(alarmed, bool):
+                raise exceptions.ValidationError('alarmed should be boolean.')
```
Comment (liusheng) [inline openlabcmd/openlabcmd/zk.py]: boolean value

File: openlabcmd/openlabcmd/cli.py (hunk 10 above)
Comment (huangtianhua) [reply openlabcmd/openlabcmd/cli.py]: I mean this should be 'Get a service'.

File: openlabcmd/openlabcmd/service.py (hunk 16 above)
Comment (huangtianhua) [reply openlabcmd/openlabcmd/service.py]: +1

File: openlabcmd/openlabcmd/cli.py [hunk 20]
```diff
@@ -124,6 +124,11 @@ def _add_ha_service_cmd(self, parser):
         cmd_ha_service_list = cmd_ha_service_subparsers.add_parser(
             'list', help='List all services.')
         cmd_ha_service_list.set_defaults(func=self.ha_service_list)
+        cmd_ha_service_list.add_argument(
+            '--node', help='Filter the services with the specified node.')
```
Comment (huangtianhua) [inline openlabcmd/openlabcmd/cli.py]: "...... specified node name" is more clear.

File: openlabcmd/openlabcmd/zk.py [hunk 21]
```diff
@@ -206,37 +217,116 @@ def delete_node(self, node_name):
...
+            if node_role_filter and exist_node['role'] not in node_role_filter:
+                continue
             path = '/ha/%s/%s_service' % (exist_node['name'],
                                           exist_node['role'])
             for service_name in self.client.get_children(path):
                 service_path = path + '/' + service_name
                 service_info = self.client.get(service_path)
-                service_info = self._bytesToDict(service_info[0])
+                service_info = self._bytesToDict(service_info,
+                                                 include_updated_at=True)
```
Comment (huangtianhua) [inline openlabcmd/openlabcmd/zk.py:254]: Just a question: when the update_at is changed for service, there is not set or update operate for service, right? if it's changed, why created_at is not needed here?

Comment (huangtianhua) [reply openlabcmd/openlabcmd/zk.py:254]: sorry, if it's can be changed.....

File: openlabcmd/openlabcmd/node.py (hunk 15 above)
Comment (wangxiyuan) [reply openlabcmd/openlabcmd/node.py]: Oh, `kwargs` is not a suitable name here. I'll change it.

File: openlabcmd/openlabcmd/zk.py (hunk 17 above)
Comment (wangxiyuan) [reply openlabcmd/openlabcmd/zk.py:41]: node_data  is a list, length is always 2.

File: openlabcmd/openlabcmd/zk.py (hunk 21 above)
Comment (huangtianhua) [reply openlabcmd/openlabcmd/zk.py:254]: sorry, misunderstand, ignore this.

File: openlabcmd/openlabcmd/node.py (hunk 13 above)
Comment (wangxiyuan) [reply openlabcmd/openlabcmd/node.py]: the type of heartbeat  is int. `heartbeat=0` means the service is just initialized.

File: openlabcmd/openlabcmd/zk.py (hunk 18 above)
Comment (wangxiyuan) [reply openlabcmd/openlabcmd/zk.py:224]: I can't see any improvement except simpler input format?

File: openlabcmd/openlabcmd/zk.py [hunk 22]
```diff
@@ -35,8 +35,22 @@ def __init__(self, config=None):
...
+        node_dict = node_data[0]
+        # ctime and mtime is millisecond in zk.
+        ctime, mtime = node_data[1].ctime, node_data[1].mtime
+
+        return_info = json.loads(node_dict.decode('utf8'))
+        if include_created_at:
+            return_info['created_at'] = datetime.datetime.fromtimestamp(
+                ctime / 1000).isoformat()
+        if include_updated_at:
+            return_info['updated_at'] = datetime.datetime.fromtimestamp(
```
Comment (kiwik) [inline openlabcmd/openlabcmd/zk.py:50]: Got it, thank you @wangxiyuan

**Review by kiwik — APPROVED** [review]
LGTM

File: openlabcmd/openlabcmd/zk.py (hunk 17 above)
Comment (liusheng) [reply openlabcmd/openlabcmd/zk.py:41]: got, thx

File: openlabcmd/openlabcmd/zk.py (hunk 18 above)
Comment (liusheng) [reply openlabcmd/openlabcmd/zk.py:224]: may you can add `action='append', default=[]` to the arguments

File: openlabcmd/openlabcmd/zk.py [hunk 23]
```diff
@@ -206,37 +217,117 @@ def delete_node(self, node_name):
...
+            role = self._get_service_role(service_name)
+        path, srv_node = self._get_service_path_and_node(service_name, role)
+        result = self.client.get(path)
+        result = self._bytesToDict(result, include_updated_at=True)
+        result['node_name'] = srv_node['name']
+        return result
+
+    @_client_check_wrapper
+    def update_service(self, service_name, role=None, alarmed=None,
+                       restarted=None, status=None, **kwargs):
```
Comment (liusheng) [inline openlabcmd/openlabcmd/zk.py:298]: didn't see any usage of the `**kwargs` argument in this function

Comment (wangxiyuan) [reply openlabcmd/openlabcmd/zk.py:298]: See L328

File: openlabcmd/openlabcmd/zk.py (hunk 18 above)
Comment (wangxiyuan) [reply openlabcmd/openlabcmd/zk.py:224]: I don't think merge different filters into once arg is a good way. It's not comfortable for user experience IMO. Take `openstack server list` for example, each filter use the arg itself.

File: openlabcmd/openlabcmd/zk.py (hunk 23 above)
Comment (liusheng) [reply openlabcmd/openlabcmd/zk.py:298]: oh, sorry.

File: openlabcmd/openlabcmd/zk.py (hunk 18 above)
Comment (wangxiyuan) [reply openlabcmd/openlabcmd/zk.py:224]: emm, I think i misunderstand your message before.  See new PS please

Comment (liusheng) [reply openlabcmd/openlabcmd/zk.py:224]: maybe you don't need L229~L240 if you added `default=[]` in the arguments

Comment (wangxiyuan) [reply openlabcmd/openlabcmd/zk.py:224]: openlabcmd is not only used for operator, but also for keepalived or some other manage tools in the future.  L229-L240 can make the input more flexible.

Comment (liusheng) [reply openlabcmd/openlabcmd/zk.py:224]: OK,

**Review by liusheng — APPROVED** [review]

**Pull request merged**